File: python/traceix_sdk.py
import os

import requests


# error classes for issues
class NoApiKey(Exception): pass


class InvalidSearchType(Exception): pass


class NoUuidProvided(Exception): pass


class TraceixSdk(object):

    """
    SDK object, simple import and call
    """

    # SDK version
    sdk_version = "0.0.0.1"

    def __init__(self, api_key=None):
        """
        initialize everything
        :param api_key: user API key, if `None` will search ENV for TRACEIX_API_KEY
        """
        if api_key is None:
            self.api_key = os.getenv("TRACEIX_API_KEY")
        else:
            self.api_key = api_key
        if self.api_key is None or self.api_key == "":
            raise NoApiKey("You did not provide an API key")
        self.base_url = "https://ai.perkinsfund.org"

    def __build_user_agent(self):
        """
        builds the user agent string to send with each request. Sends OS and Python version
        :return: str
        """
        # set the env var TRACEIX_DISABLE_TELEMETRY to 1 to disable telemetry data
        telemetry_disabled = os.environ.get("TRACEIX_DISABLE_TELEMETRY") == "1"
        user_agent = f"Traceix/{self.sdk_version}"
        if not telemetry_disabled:
            import platform

            user_agent += f" ({platform.platform()} v{platform.python_version()})"
        return user_agent

    def __build_headers(self):
        """
        builds the headers to send with each request
        :return: dict
        """
        return {
            "x-api-key": self.api_key,
            "user-agent": self.__build_user_agent()
        }

    def __build_url(self, path):
        """
        builds the full URL to send the request to
        :param path: the URL path
        :return: str
        """
        return f"{self.base_url}/{path}"

    @staticmethod
    def __build_file_data(filename):
        """
        builds file POST data to send with the request
        :param filename: the full path to the file
        :return: dict
        """
        fh = open(filename, "rb")
        files = {
            "file": (filename, fh, "application/octet-stream")
        }
        return files

    def full_upload(self, filename):
        """
        performs 3 requests to make a full upload of the file
        :param filename: the full path to the file
        :return: tuple
        """
        ai_data = self.ai_prediction(filename)
        capa_status = self.capa_extraction(filename)
        exif_data = self.exif_extraction(filename)
        return ai_data, capa_status, exif_data

    def ai_prediction(self, filename):
        """
        sends a request to the prediction endpoint
        :param filename: full path to the file
        :return: dict
        """
        url = self.__build_url("/api/traceix/v1/upload")
        headers = self.__build_headers()
        post_data = self.__build_file_data(filename)
        try:
            req = requests.post(url, headers=headers, files=post_data)
        except:
            req = None
        if req is not None:
            return req.json()
        else:
            return None

    def check_status(self, uuid):
        """
        check the status of a provided UUID
        :param uuid: the UUID string
        :return: dict
        """
        if uuid is None:
            raise NoUuidProvided("You did not provide a UUID required by the endpoint")
        url = self.__build_url("/api/v1/traceix/status")
        headers = self.__build_headers()
        post_data = {"uuid": uuid}
        try:
            req = requests.post(url, headers=headers, json=post_data)
        except:
            req = None
        if req is not None:
            return req.json()
        else:
            return None

    def hash_search(self, file_hash, **kwargs):
        """
        search by file hash on the provided endpoints to get the associated data
        :param file_hash: the hash of the file
        :param kwargs: search_type - the type of search either capa or exif (**default capa)
        :return: dict
        """
        search_type = kwargs.get("search_type", "capa")

        if search_type == "capa":
            url = self.__build_url("/api/traceix/v1/capa/search")
        elif search_type == "exif":
            url = self.__build_url("/api/traceix/v1/exif/search")
        else:
            url = None
        if url is None:
            raise InvalidSearchType("Search must be of type capa or exif")
        headers = self.__build_headers()
        post_data = {"sha256": file_hash}
        headers["content-type"] = "application/json"
        try:
            req = requests.post(url, headers=headers, json=post_data)
        except:
            req = None
        if req is not None:
            return req.json()
        else:
            return None

    def capa_extraction(self, filename):
        """
        extract the capabilities from the filename
        :param filename: the full path to the file
        :return: dict
        """
        url = self.__build_url("/api/traceix/v1/capa")
        headers = self.__build_headers()
        post_data = self.__build_file_data(filename)
        try:
            req = requests.post(url, headers=headers, files=post_data)
        except:
            req = None
        if req is not None:
            return req.json()
        else:
            return None

    def exif_extraction(self, filename):
        """
        extract the exif metadata from the filename
        :param filename: the full path to the file
        :return: dict
        """
        url = self.__build_url("/api/traceix/v1/exif")
        headers = self.__build_headers()
        post_data = self.__build_file_data(filename)
        try:
            req = requests.post(url, headers=headers, files=post_data)
        except:
            req = None
        if req is not None:
            return req.json()
        else:
            return None

    def list_all_ipfs_datasets(self):
        """
        list all the public datasets available currently
        NOTE: API key is not needed for these endpoints
        :return: dict
        """
        url = self.__build_url("/api/traceix/v1/ipfs/listall")
        headers = self.__build_headers()
        try:
            req = requests.post(url, headers=headers)
        except:
            req = None
        if req is not None:
            return req.json()
        else:
            return None

    def get_public_ipfs_dataset(self, cid):
        """
        get the public dataset by the CID
        :param cid: CID string
        :return: dict
        """
        url = self.__build_url("/api/traceix/v1/ipfs/search")
        headers = self.__build_headers()
        post_data = {
            "cid": cid
        }
        try:
            req = requests.post(url, headers=headers, json=post_data)
        except:
            req = None
        if req is not None:
            return req.json()
        else:
            return None

    def search_ipfs_dataset_by_hash(self, file_hash):
        """
        search by file hash to see if the dataset has been uploaded to public domain
        :param file_hash: hash string
        :return: dict
        """
        url = self.__build_url("/api/traceix/v1/ipfs/find")
        headers = self.__build_headers()
        post_data = {
            "sha_hash": file_hash
        }
        try:
            req = requests.post(url, headers=headers, json=post_data)
        except:
            req = None
        if req is not None:
            return req.json()
        else:
            return None

File: javascript/traceix-sdk.js
const fs = require("fs");
const os = require("os");
const axios = require("axios");
const FormData = require("form-data");

// Error classes
class NoApiKey extends Error {
  constructor(message = "You did not provide an API key") {
    super(message);
    this.name = "NoApiKey";
  }
}

class InvalidSearchType extends Error {
  constructor(message = "Search must be of type capa or exif") {
    super(message);
    this.name = "InvalidSearchType";
  }
}

class NoUuidProvided extends Error {
  constructor(message = "You did not provide a UUID required by the endpoint") {
    super(message);
    this.name = "NoUuidProvided";
  }
}

class TraceixSdk {
  /**
   * SDK object, simple import and call
   */
  static sdkVersion = "0.0.0.1";

  /**
   * Initialize everything
   * @param {string|null} apiKey - user API key, if null will search ENV for TRACEIX_API_KEY
   */
  constructor(apiKey = null) {
    if (apiKey == null) {
      this.apiKey = process.env.TRACEIX_API_KEY || "";
    } else {
      this.apiKey = apiKey;
    }

    if (!this.apiKey || this.apiKey === "") {
      throw new NoApiKey();
    }

    this.baseUrl = "https://ai.perkinsfund.org";
  }

  /**
   * Builds the user agent string to send with each request.
   * Sends OS and Node version unless TRACEIX_DISABLE_TELEMETRY=1
   * @return {string}
   * @private
   */
  _buildUserAgent() {
    const telemetryDisabled = process.env.TRACEIX_DISABLE_TELEMETRY === "1";
    let userAgent = `Traceix/${TraceixSdk.sdkVersion}`;

    if (!telemetryDisabled) {
      const platform = `${os.type()} ${os.release()}`;
      const nodeVersion = process.version;
      userAgent += ` (${platform} v${nodeVersion})`;
    }

    return userAgent;
  }

  /**
   * Builds the headers to send with each request
   * @return {Object}
   * @private
   */
  _buildHeaders() {
    return {
      "x-api-key": this.apiKey,
      "user-agent": this._buildUserAgent(),
    };
  }

  /**
   * Builds the full URL to send the request to
   * @param {string} path
   * @return {string}
   * @private
   */
  _buildUrl(path) {
    if (path.startsWith("/")) {
      return `${this.baseUrl}${path}`;
    }
    return `${this.baseUrl}/${path}`;
  }

  /**
   * Builds file POST data to send with the request
   * @param {string} filename
   * @return {FormData}
   * @private
   */
  _buildFileData(filename) {
    const form = new FormData();
    form.append("file", fs.createReadStream(filename), {
      filename,
      contentType: "application/octet-stream",
    });
    return form;
  }

  /**
   * Performs 3 requests to make a full upload of the file
   * @param {string} filename - full path to the file
   * @return {Promise<[any, any, any]>}
   */
  async fullUpload(filename) {
    const aiData = await this.aiPrediction(filename);
    const capaStatus = await this.capaExtraction(filename);
    const exifData = await this.exifExtraction(filename);
    return [aiData, capaStatus, exifData];
  }

  /**
   * Sends a request to the prediction endpoint
   * @param {string} filename - full path to the file
   * @return {Promise<any|null>}
   */
  async aiPrediction(filename) {
    const url = this._buildUrl("/api/traceix/v1/upload");
    const headers = this._buildHeaders();
    const form = this._buildFileData(filename);

    try {
      const resp = await axios.post(url, form, {
        headers: {
          ...headers,
          ...form.getHeaders(),
        },
      });
      return resp.data;
    } catch (e) {
      return null;
    }
  }

  /**
   * Check the status of a provided UUID
   * @param {string|null} uuid
   * @return {Promise<any|null>}
   */
  async checkStatus(uuid) {
    if (uuid == null) {
      throw new NoUuidProvided();
    }

    const url = this._buildUrl("/api/v1/traceix/status");
    const headers = this._buildHeaders();
    const postData = { uuid };

    try {
      const resp = await axios.post(url, postData, {
        headers: {
          ...headers,
          "content-type": "application/json",
        },
      });
      return resp.data;
    } catch (e) {
      return null;
    }
  }

  /**
   * Search by file hash on the provided endpoints to get the associated data
   * @param {string} fileHash - the hash of the file
   * @param {Object} options - { search_type: "capa" | "exif" }
   * @return {Promise<any|null>}
   */
  async hashSearch(fileHash, options = {}) {
    const searchType = options.search_type || "capa";
    let url;

    if (searchType === "capa") {
      url = this._buildUrl("/api/traceix/v1/capa/search");
    } else if (searchType === "exif") {
      url = this._buildUrl("/api/traceix/v1/exif/search");
    } else {
      url = null;
    }

    if (!url) {
      throw new InvalidSearchType();
    }

    const headers = this._buildHeaders();
    const postData = { sha256: fileHash };

    try {
      const resp = await axios.post(url, postData, {
        headers: {
          ...headers,
          "content-type": "application/json",
        },
      });
      return resp.data;
    } catch (e) {
      return null;
    }
  }

  /**
   * Extract the capabilities from the filename
   * @param {string} filename
   * @return {Promise<any|null>}
   */
  async capaExtraction(filename) {
    const url = this._buildUrl("/api/traceix/v1/capa");
    const headers = this._buildHeaders();
    const form = this._buildFileData(filename);

    try {
      const resp = await axios.post(url, form, {
        headers: {
          ...headers,
          ...form.getHeaders(),
        },
      });
      return resp.data;
    } catch (e) {
      return null;
    }
  }

  /**
   * Extract the exif metadata from the filename
   * @param {string} filename
   * @return {Promise<any|null>}
   */
  async exifExtraction(filename) {
    const url = this._buildUrl("/api/traceix/v1/exif");
    const headers = this._buildHeaders();
    const form = this._buildFileData(filename);

    try {
      const resp = await axios.post(url, form, {
        headers: {
          ...headers,
          ...form.getHeaders(),
        },
      });
      return resp.data;
    } catch (e) {
      return null;
    }
  }

  /**
   * List all the public datasets available currently
   * NOTE: API key is not needed for these endpoints (but we still send headers, like Python)
   * @return {Promise<any|null>}
   */
  async listAllIpfsDatasets() {
    const url = this._buildUrl("/api/traceix/v1/ipfs/listall");
    const headers = this._buildHeaders();

    try {
      const resp = await axios.post(url, null, { headers });
      return resp.data;
    } catch (e) {
      return null;
    }
  }

  /**
   * Get the public dataset by the CID
   * @param {string} cid
   * @return {Promise<any|null>}
   */
  async getPublicIpfsDataset(cid) {
    const url = this._buildUrl("/api/traceix/v1/ipfs/search");
    const headers = this._buildHeaders();
    const postData = { cid };

    try {
      const resp = await axios.post(url, postData, {
        headers: {
          ...headers,
          "content-type": "application/json",
        },
      });
      return resp.data;
    } catch (e) {
      return null;
    }
  }

  /**
   * Search by file hash to see if the dataset has been uploaded to public domain
   * @param {string} fileHash
   * @return {Promise<any|null>}
   */
  async searchIpfsDatasetByHash(fileHash) {
    const url = this._buildUrl("/api/traceix/v1/ipfs/find");
    const headers = this._buildHeaders();
    const postData = { sha_hash: fileHash };

    try {
      const resp = await axios.post(url, postData, {
        headers: {
          ...headers,
          "content-type": "application/json",
        },
      });
      return resp.data;
    } catch (e) {
      return null;
    }
  }
}

module.exports = {
  TraceixSdk,
  NoApiKey,
  InvalidSearchType,
  NoUuidProvided,
};
